micro_dl/utils/image_utils.py

```diff
@@ -190,3 +219,69 @@ def read_image(file_path):
         except IOError as e:
             raise e
     return im
+
+
+def get_unimodal_threshold(input_image):
+    """Determines optimal unimodal threshold
+
+    https://users.cs.cf.ac.uk/Paul.Rosin/resources/papers/unimodal2.pdf
+    https://www.mathworks.com/matlabcentral/fileexchange/45443-rosin-thresholding
+
+    :param np.array input_image: generate mask for this image
+    :return float best_threshold: optimal lower threshold for the fg hist
+    """
+
+    hist_counts, bin_edges = np.histogram(
+        input_image,
+        bins=256,
+        range=(input_image.min(), np.percentile(input_image, 99.5))
+    )
+    bin_centers = (bin_edges[:-1] + bin_edges[1:]) / 2
+
+    max_idx = np.argmax(hist_counts)
+    int_with_max_count = bin_centers[max_idx]
+    p1 = [int_with_max_count, hist_counts[max_idx]]
+
+    # find last non-empty bin
+    pos_counts_idx = np.where(hist_counts > 0)[0]
+    last_binedge = pos_counts_idx[-1]
+    p2 = [bin_centers[last_binedge], hist_counts[last_binedge]]
+
+    best_threshold = -1
+    max_dist = -1
+    for idx in range(max_idx, last_binedge, 1):
+        x0 = bin_centers[idx]
+        y0 = hist_counts[idx]
+        a = [p1[0] - p2[0], p1[1] - p2[1]]
+        b = [x0 - p2[0], y0 - p2[1]]
+        cross_ab = a[0] * b[1] - b[0] * a[1]
+        per_dist = np.linalg.norm(cross_ab) / np.linalg.norm(a)
+        if per_dist > max_dist:
+            best_threshold = x0
+            max_dist = per_dist
+
+    return best_threshold
```

Comment:
Would there ever be a histogram where best_threshold stays -1 (e.g. max_idx == last_binedge)? If so, maybe do a final check that best_threshold is > -1?

Reply:
good point! The images are uint16 and not normalized. After flat field correction, they are float but with non negative values.  Changing best_threshold = np.inf, max_dist = -np.inf and adding an assert before return!